New feature docs

greg-mckeon opened the pull request:
There's more to add here, but starting the PR off.

greg-mckeon committed:
add durable objects pages. rerank pages.

greg-mckeon committed:
Add usage guide

koeninger commented on src/content/learning/how-durable-objects-work.md:
```diff
@@ -0,0 +1,216 @@
+
+Durable Objects are single-threaded, but each Object has almost no overhead associated with it.  In general, this means your application should look to create and access many Durable Objects over sending many requests to the same Object.  Access patterns that make many requests to a single Object will see degraded performance.
+
+Since Durable Objects are globally distributed, initial creation latency may be high when using system-generated ids (see above).  Until automatic migration of Durable Objects is implemented, latency will also be high on subsequent read and write operations.
```
Is this intended to say client-defined ids?

koeninger commented on src/content/learning/how-durable-objects-work.md:
```diff
@@ -0,0 +1,216 @@
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  
+
+During a network partition between data centers or between machines within a data center, it is possible for more than one Durable Object with a given id to be active.  While this is unlikely, we plan to do additional work to reduce the chance of a network partition during the beta period.
```
What are we planning to do to reduce the chance of a network partition?

a-robinson replied:
Ignore this, network partitions can't cause multiple durable objects with the same ID to be receiving requests at the same time. This adds nothing to the previous paragraph other than confusion.

greg-mckeon replied:
For my own understanding, this is because the a durable object will need to write to the ownership table, which it won't be able to do if there's a partition?

koeninger commented on src/content/learning/how-durable-objects-work.md:
```diff
@@ -0,0 +1,216 @@
+
+### Global Uniqueness
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  
```
I'd explicitly call out that this means when a script is updated, existing requests may still be running on the old object concurrently with new requests going to the new object.

greg-mckeon committed:
rename, fix typo + ordering

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+* Global Uniqueness guarantees that there will be a single Durable Object with a given id running at once.  Requests for a Durable Object id are routed by the Workers runtime to the Cloudflare point-of-presence that owns the Durable Object.  
+
+* The transactional storage API provides strongly-consistent, key-value storage to the Durable Object.  Each Object can only read and modify keys associated with that Object. Execution of a Durable Object is single-threaded, but multiple request threads may be processed out-of-order from how they arrived at the Object.
```
> but multiple request threads may be processed out-of-order from how they arrived at the Object.

The wording is here is a little strange. Maybe switch it to "multiple request events" instead of threads

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+To define a Durable Object, developers export an ordinary Javascript class.  Other languages will need a shim that translates their class definition to a Javascript class.
+
+The first variable passed to the class constructor contains state specific to the Durable Object, including methods for accessing storage.
```
The natural next question that doesn't appear to be answered here is what's the second variable passed to the constructor?

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+    constructor(state, environment){
+    }
+    
+    addEventListener("fetch", (event) => {
```
`addEventListener` isn't used in class-based actors. There just needs to be a function with the prototype `async fetch(request) {`

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+### Accessing Storage from a Durable Object
+
+Durable Objects gain access to a TODO: fill in[persistent storage API]() via the first parameter passed to the Durable Object constructor.  While access to a Durable Object is single-threaded, it's important to remember that there can still be race conditions across multiple requests.
```
> it's important to remember that there can still be race conditions across multiple requests.

I'd reword this to something like "it's important to remember that request executions can still interleave with each other when they wait on I/O, such as when waiting on the promises returned by persistent storage methods or `fetch` requests."

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
```
Again, get rid of this and change the name of the function below from `handle` to `fetch`.

a-robinson commented on src/content/learning/using-durable-objects.md:
````diff
@@ -0,0 +1,216 @@
+
+```
+
+Single statement transactions are always transactional.  More complex use cases can wrap multiple storage statements in a transaction.  
````
"Single statement transactions are always transactional" comes off sounding a bit strange. Perhaps "Each individual storage operation behaves like a database transaction."

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
```
Again, get rid of this and change the name of the function below from `handle` to `fetch`.

a-robinson commented on src/content/learning/using-durable-objects.md:
````diff
@@ -0,0 +1,216 @@
+
+```js
+
+class DurableObjectExample {
````
It'd be nice to name each of these class differently for each example as a way of indicating what the example is meant to show.

a-robinson commented on src/content/learning/using-durable-objects.md:
````diff
@@ -0,0 +1,216 @@
+
+```
+
+Transactions can fail if they conflict with a concurrent transaction from the same Durable Object.  Transactions are transparently and automatically retried once before returning a failure.
````
I'd insert an extra little clause here, changing "retried once before" to "retried once by rerunning the provided function before"

a-robinson replied:
This could also be a place to add in that this is because all transactions are run at serializable isolation, making them strongly consistent. Totally optional though

a-robinson commented on src/content/learning/using-durable-objects.md:
````diff
@@ -0,0 +1,216 @@
+First, upload the script that contains the class via the Workers Dashboard or via the command line.
+
+```sh
+curl -i -H "X-Auth-Email: ${EMAIL}" -H "X-Auth-Key: ${AUTH_KEY}" "https://api.cloudflare.com/client/v4/accounts/${ACCOUNT_TAG}/workers/scripts/${SCRIPT_NAME}" -X PUT -H "Content-Type: application/javascript" --data "`cat durable_object_example.js`"
````
We might encourage people to use API tokens rather than auth keys, which just requires a header like "Authorization: Bearer <token>". Probably worth checking out what other parts of the docs use.

a-robinson commented on src/content/learning/using-durable-objects.md:
````diff
@@ -0,0 +1,216 @@
+To create the class, call the following endpoint.
+
+```sh
+curl -i -H "X-Auth-Email: ${EMAIL}" -H "X-Auth-Key: ${AUTH_KEY}" "https://api.cloudflare.com/client/v4/accounts/${ACCOUNT_TAG}/workers/durable_objects/namespaces" -X POST --data "{'name': 'example-class', 'script': ${SCRIPT_NAME}"
````
The JSON blob appears syntactically invalid, and also a class name needs to be specified  as well. Use this instead:

```
"{'name': 'example-class', 'script': 'example-script', 'class: 'example-class'}"
```

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+  "body_part": "script",
+  "bindings": [
+    {
+      "type": "durable_object_namespace",
```
Needs to be "durable_object_class"

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+    {
+      "type": "durable_object_namespace",
+      "name": "example-class",
+      "namespace_id": "<namespace id>"
```
Needs to be "class_id" instead of "namespace_id"

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+ 
+async function handleRequest(request) {
+  // Retrieve the Fetcher for the Durable Object with id 1
+  let durableObject = await DURABLE_OBJECT_CLASS.get("1")
```
It'd be nice to make this binding name ("DURABLE_OBJECT_CLASS") match the "name" provided in the binding definition in the section above, since that's where it comes from.

a-robinson replied:
This isn't how requesting an actor works. You first need to get an ID object, then pass it to the durable object class binding, like this:

```
let objectId = DURABLE_OBJECT_CLASS.idFromName("1")
let durableObject = await DURABLE_OBJECT_CLASS.get(objectId)
```

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+  // Retrieve the Fetcher for the Durable Object with id 1
+  let durableObject = await DURABLE_OBJECT_CLASS.get("1")
+  // Send a request to Durable Object '1', with the URI "get-count"
+  let resp = await durableObject.fetch("get-count")
```
All URIs need to be validly formed, including either an `http://` or `https://` prefix. We'll likely remove this restriction in the future, but it's important for now. So change this to something like `"https://get-count"`

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+  let count = await resp.text()
+ 
+  // Regardless of the URI, requests are always routed to the Durable Object
+  resp = await durableObject.fetch()
```
A fetch without a URL at all definitely won't work.

Before publishing we'll need to make sure to run these examples for real and make sure they all work.

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+  let message = await resp.text()
+
+  // TODO: system generated id example
+  let id = generateId()
```
```
  let systemGeneratedId = DURABLE_OBJECT_CLASS.newUniqueId()
  let sysDurableObject = await DURABLE_OBJECT_CLASS.get(systemGeneratedId)
  resp = await sysDurableObject.fetch('https://get-count')
```

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+### Global Uniqueness
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  
```
> In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  

If we want to be more specific, at the present this only time during which you can currently have overlap of an old instance of a durable object and a new instance is when pushing a new version of your code. It's understandable if we don't want to be that specific though.

Along those lines, though, it's worth including an example that shows using the class's in-memory state for something, and then perhaps calling out down here that in-memory state is transient and can disappear when an actor needs to be stopped or replaced, such as when a new its code is pushed or when it doesn't receive traffic for some time between requests.

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  
+
+During a network partition between data centers or between machines within a data center, it is possible for more than one Durable Object with a given id to be active.  While this is unlikely, we plan to do additional work to reduce the chance of a network partition during the beta period.
```
This is not true. The only sense in which it's possible for multiple instances of a durable object to be active at the same is that requests to an old instance might still be executing when the new instance starts receiving requests. It is explicitly **not** possible for a partition to cause two different instances of an object to be receiving requests at the same point in time.

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+### Creation and Routing
+
+Durable Objects use a deterministic hash of the Object's id to determine the Cloudflare point-of-presence that owns the Object.  It is not currently possible for an Object to migrate ownership to a different point-of-presence.
```
> Durable Objects use a deterministic hash of the Object's id to determine the Cloudflare point-of-presence that owns the Object.

We use what's effectively a hash for durable object IDs generated using `ns.idFromName('some-string')`. We always put durable objects created using a system-generated ID in the closest possible point-of-presence. I see that your next sentence covers system generated IDs, but this sentence should not exist, at least not without making clear that it's only true for client-defined IDs.

Honestly, I'd just remove these two sentences, and put the sentence about how objects can not currently be migrated to different parts of the world before what's currently the last sentence in this section.

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+### Transactional Storage API
+
+The transactional storage API is scoped to a single Durable Object.  It is not currently possible to access data stored in a Durable Object from a different Durable Object directly through the API.
```
> It is not currently possible to access data stored in a Durable Object from a different Durable Object directly through the API

Minor nit, but I'd reword this a little to be:
"It is not currently possible to access data stored in one Durable Object from a different Durable Object."

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+The transactional storage API is scoped to a single Durable Object.  It is not currently possible to access data stored in a Durable Object from a different Durable Object directly through the API.
+
+There is no support for querying, bulk loading, or exporting data stored in Durable Objects.
```
"querying" is such a vague word, and it's not that it isn't possible to load or export data so much as that you have to write your own code to do it for now. If you don't mind all the edits, I'd make this sentence:

"There is no support for advanced queries or bulk loading or exporting of data. Loading or exporting data can be done by writing custom Workers and Durable Objects to do so yourself.

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+There is no support for querying, bulk loading, or exporting data stored in Durable Objects.
+
+Applications should be prepared for transactions to fail due to conflicts, and to retry those failed transactions.  A failed transaction will throw an error (TODO: what error?  is this true?) that can be caught.  It is always safe to retry failed transactions.
```
I'd say "Applications that use multi-operation transactions ..." to be clear.

a-robinson replied:
The error thrown will include text saying "Transaction failed due to conflict".

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,216 @@
+
+Durable Objects are single-threaded, but each Object has almost no overhead associated with it.  In general, this means your application should look to create and access many Durable Objects over sending many requests to the same Object.  Access patterns that make many requests to a single Object will see degraded performance.
+
+Since Durable Objects are globally distributed, initial creation latency may be high when using user-generated ids (see above).  Until automatic migration of Durable Objects is implemented, latency will also be high on subsequent read and write operations.
```
s/user-generated/user-provided

a-robinson replied:
> latency will also be high on subsequent read and write operations.

"latency will also be high on subsequent requests to such Durable Objects".

a-robinson reviewed (commented):
Sorry for the big dump of comments. Are you gonna need some help finishing these up?

greg-mckeon committed:
add initial runtime apis

kentonv committed:
Various Durable Objects updates.

greg-mckeon committed:
address a few comments

a-robinson committed:
Edits and minor additions to Using Durable Objects page

a-robinson commented:
Pushed updates to resolve my remaining comments that Kenton's changes didn't already handle.

kentonv committed:
Reference docs for durable object namespace bindings.

kentonv commented:
Pushed some more stuff. I realize after the fact that I didn't follow the usual pattern for documenting class methods when I documented bindings, but I'm out of time.

I started calling the bindings "namespaces" because calling them "class bindings" was getting too confusing. I kind of think we may want to change the API back to `/workers/durable_objects/namespaces` rather than `/classes`... you configure a namespace that uses a particular class. Calling the namespace itself a "class" leads to too much confusion when trying to explain them.

Another, wilder idea: Maybe we shouldn't make people configure the namespaces separately at all. Maybe they could be implicitly created when you upload the script. (This idea obviously has lots of issues to work through but wanted to plant the thought to chew on...)

a-robinson committed:
Rename "controller" to "state" and "environment" to "env" in new docs

Since "state" and "env" appear to be what we're calling them in the API
Reference docs, and we should try to be consistent.

greg-mckeon committed:
merging

greg-mckeon commented:
Not merging just yet, want to give @koeninger a chance to look.  Then will add a separate PR for documenting the API specifically.

koeninger commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,302 @@
+
+## Limitations during the Beta
+
+<<<<<<< HEAD
```
Just want to make sure this conflict is handled before merging

koeninger commented:
I'm still in favor of just linking to the shell script rather than all of the details about the upload api.

greg-mckeon committed:
add example code

greg-mckeon committed:
add upload script

a-robinson committed:
Fix / flesh out storage runtime API documentation

koeninger commented:
Not a big deal, but I think it should be possible to add the shell script as a static file and link it rather than copy-paste, e.g. https://www.gatsbyjs.com/docs/static-folder/

But this seems fine for now.

a-robinson committed:
s/Durable Object classes/Durable Object namespaces

To avoid confusion between the JS class that implements a durable
namespace and the higher-level namespace of durable objects (which can
be reconfigured to use a different JS class as its implementation).

a-robinson committed:
Remove hanging reference to chat demo in docs

greg-mckeon committed:
add static script

greg-mckeon committed:
replace some instances of class w/namespace

greg-mckeon committed:
add link to shell script

a-robinson commented on src/content/learning/using-durable-objects.md:
```diff
@@ -0,0 +1,369 @@
+
+## Configuration Script
+
+To simplify steps
```
Left this sentence hanging? I assume it's meant to link to the script?

greg-mckeon committed:
1update shell script link and add beta warning

a-robinson committed:
Handful of fixes and small edits for durable object docs

a-robinson committed:
Document the rollback() method on Durable Object transactions

greg-mckeon committed:
fixed bugs so examples actually work

greg-mckeon committed:
merged Alex's changes

greg-mckeon commented:
Add Durable Objects docs
